oculus_ros2/include/oculus_ros2/sonar_viewer.hpp

```diff
@@ -0,0 +1,138 @@
+// Copyright 2023 Forssea Robotics
+// All rights reserved.
+//
+// Unauthorized copying of this code base via any medium is strictly prohibited.
+// Proprietary and confidential.
+
+#ifndef OCULUS_ROS2__SONAR_VIEWER_HPP_
+#define OCULUS_ROS2__SONAR_VIEWER_HPP_
+
+#include <cv_bridge/cv_bridge.h>
+#include <oculus_driver/AsyncService.h>
+#include <oculus_driver/SonarDriver.h>
+
+#include <algorithm>
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include <oculus_interfaces/msg/ping.hpp>
+#include <oculus_ros2/conversions.hpp>
+#include <opencv2/core.hpp>
+#include <opencv2/highgui/highgui.hpp>
+#include <opencv2/imgproc/imgproc.hpp>
+#include <rclcpp/rclcpp.hpp>
+#include <sensor_msgs/msg/image.hpp>
+#include <std_msgs/msg/header.hpp>
+
+class SonarViewer {
+public:
+  // SonarViewer(bool *arg);
+  explicit SonarViewer(rclcpp::Node* node);
+  ~SonarViewer();
+  void publish_fan(const oculus::PingMessage::ConstPtr& ping, const std::string& frame_id = "", const int& data_depth = 0) const;
+  void publish_fan(
+      const oculus_interfaces::msg::Ping& ros_ping_msg, const std::string& frame_id = "", const int& data_depth = 0) const;
+  template <typename dataType>
+  void publish_fan(const int& width,
+      const int& height,
+      const int& offset,
+      const std::vector<dataType>& ping_data,
+      const int& master_mode,
+      const double& ping_rage,
+      const std_msgs::msg::Header& header) const;
+  void stream_and_filter(const oculus::PingMessage::ConstPtr& ping, cv::Mat& data);
+  // void stream_and_filter(const oculus::PingMessage::ConstPtr &ping);
+  // sensor_msgs::msg::Image publish_fan(const oculus::PingMessage::ConstPtr &ping);
+
+private:
+  //      //   sensor_msgs::ImagePtr msg;
+  const rclcpp::Node* node_;
+  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr image_publisher_;
+};
+
+template <typename T>
```

Comment:
This is better : [https://stackoverflow.com/questions/55102504/populate-a-vector-with-linearly-increased-values](https://stackoverflow.com/questions/55102504/populate-a-vector-with-linearly-increased-values)

Reply:
Like that :
```C++
template <typename T>
std::vector<double> linspace(T start_in, T end_in, int num_in) {
  const auto start = static_cast<double>(start_in);
  const auto end = static_cast<double>(end_in);
  const auto num = static_cast<double>(num_in);
  std::vector<double> linspaced(std::max(0,num_in-1));

  if (num == 0) {
    return linspaced;
  }
  if (num == 1) {
    linspaced.push_back(start);
    return linspaced;
  }

  double delta = (end - start) / (num - 1);
  std::generate(linspaced.begin(), linspaced.end(), [i = 0, &delta]() mutable { return i++ * delta; });
  linspaced.push_back(end);  // I want to ensure that start and end
                             // are exactly the same as the input
  return linspaced;
}
```
?